src/test/java/org/candlepin/controller/PoolManagerFunctionalTest.java

```diff
@@ -1560,14 +1448,130 @@ public void testEntitlementQuantityChange() throws EntitlementRefusedException {
         List<Entitlement> entitlements = poolManager.entitleByPools(consumer, poolQuantities);
 
         Entitlement ent = entitlements.get(0);
-        assertTrue(ent.getQuantity() == 3);
+        assertEquals(3, (int) ent.getQuantity());
 
         poolManager.adjustEntitlementQuantity(consumer, ent, 5);
         Entitlement ent2 = entitlementCurator.get(ent.getId());
-        assertTrue(ent2.getQuantity() == 5);
+        assertEquals(5, (int) ent2.getQuantity());
 
         Pool pool2 = poolCurator.get(consumerPool.getId());
-        assertTrue(pool2.getConsumed() == 5);
-        assertTrue(pool2.getEntitlements().size() == 1);
+        assertEquals(5, (long) pool2.getConsumed());
+        assertEquals(1, pool2.getEntitlements().size());
     }
+
+    private static class MockSubscriptionServiceAdapter implements SubscriptionServiceAdapter {
```

Comment:
Why move this declaration to the bottom of the class? If we're going to standardize on this placement (and I would argue we *should*), following C-style declare-before-use makes for much more readable code overall IMO. By the time you get to any particular line, you've had the chance to see the code whatever you're currently reading is reliant upon.

Reply:
I would argue for the opposite. We you enter a class you should see the important stuff first. Public methods and such. Private methods and nested classes are implementation details. They are in general case irrelevant and best put somewhere (like at the bottom of the file) where they don't obscure actually important pieces of code.

In C you don't have a choice. You have to declare before use. Java gives you options for a more natural narrative throughout the code instead of forcing you to follow arbitrary technical limitations of legacy languages.

Reply:
When you're looking at the source files directly, it's no longer about high-level details -- you're literally reading the implementation at that point. And when reading an implementation, the fewer jumps around the code you can impart on the reader, the better.

If I'm reading through code, especially for a review or evaluating a file from top to bottom, the fewer times I need to jump around to understand the full context of a given operation, the better. The older languages technical limitation also happen to impart a natural discovery of every possible invocation (within the file) by the time you got to any bit of code. The reader has the context and knows where to get it. If they haven't seen it, it's come from another source file. Natural flows and syntax inform where things live without needing to rely on specific functionality of an IDE or educated guesses; as well as prepare the reader for what they're going to be evaluating before they get to the point where it's needed.

```
void publicFunc() {
   // ...

   result = helperFunc();
   // at this point, the reader must now stop what they're doing to jump further ahead
   // into the file to read what helperFunc is doing, before returning back to this
   // location

   // ...
}

static void* helperFunc() {
   // ...
}
```

vs

```
static void* helperFunc() {
   // ...
}

void publicFunc() {
   // ...

   result = helperFunc();
   // the reader has already had the chance to read what helperFunc is and does, and
   // if the knowledge was retained, they do not need to jump out of this function to
   // get additional context

   // ...
}
```

Reply:
In the past this was a formatting decision that was decided for me, so I had to adopt their style. But I've tried thinking through this to find a style that makes sense to me.

A class reads top to bottom and in my opinion,  I think it should start from general to more specific. So, having your public scoped method before the private method makes sense to me as the private methods hold more specific details. However, if one function calls the other, I think that that they should be vertically close if possible. I think it is a more natural flow and the reader can expect the function definition to be closely below where it is used, so they can quickly jump to the private method and get more details on the operation the function provides if they need. 

I get the point that Chris makes about reading the private methods first and understanding them so that you have the contextual knowledge first before you see how the methods are used, but I think this can be an overload of information on the reader before they have a more general understanding of the operations.  I don't think the reader needs the contextual information of the dependent functions upfront because proper naming of the dependent function should give the reader enough high level contextual information to understand the current method's logic without having to make that jump. 

So I think this makes sense to me:
```java
import com.blah.whatever

public class MyClass {
    private static final int MY_INT = 1000;

    private String var = "asdf"

    public func1() {
        helpFunc1()
     }

    public func2() {
        helpFunc1()
     }

    private helpFunc1() {
    }

    public func3() {
         helpFunc2()
    }
   
    private helpFunc2() {
    }

    public func4 () {
    }
}
```

Reply:
> What you are describing are practices from C where they have them not because of choice but because of legacy limitations. The idea that if a function calls a piece of code and you haven't seen it above so it must be coming from another source file is just wrong. It's something you cannot count on in any language besides C and the like.

This is somewhat inaccurate. In C this discussion still occurred because the *declaration* were all that needed to come first. The actual implementations could be all over the place. And whether or not this situation *can* occur in other languages/projects/contexts doesn't mean it *should* occur in ours. Especially in a case like this where a change is being made to existing code rather than it being brought in net-new.

If we're willing to make ancillary changes in a PR for the sake of performing stylistic/readability changes, then then it's time to have the discussion about what is most universally readable for the development environments or flows that we encourage and support.

> irrelevant code
> insignificant implementation detail 

I feel these statements are at the crux of the issue here. If you're reading the code at all, it's because you're analyzing it in some fashion; maybe as part of some debug-adjacent static analysis, or you're doing as I'm doing now and *literally evaluating the implementation details as part of a code review*. I cannot see any instance where you're reading the code and the implementation of said code is insignificant or irrelevant. If it is, then I would argue the implementation itself has bloat issues (which, amusingly enough, are very stereotypical *Java*). The reason you'd perform an abstraction is to improve code readability and maintenance. If it's doing neither of those things, then the problem isn't where the code lives in a given file...

>  I don't think the reader needs the contextual information of the dependent functions upfront because proper naming of the dependent function should give the reader enough high level contextual information to understand the current method's logic without having to make that jump.

To an extent. When it comes to code evaluation, there's more than just the name of the symbols in play. That's a very critical aspect of it, but you also need to know what kind of abstraction you're working with: is it isolated/stateless? what state is shared? what types/data are passed in? etc..
To some degree, this gives a lot of weight to your point: you can't really fully evaluate those methods without the context in which they're called. However, the larger point is that in either case, when it comes to evaluating a given invocation, I still need to understand what it's doing at that point regardless of where it lives in the code. If I've already read it (or had the chance to and skimmed/skipped), I'm somewhat prepared and may not need to jump back up to re-read it again with the new context. Sure, it may be complex enough that I still need to regardless, but that situation won't be changed or improved by removing the choice entirely by moving it later in the file.

There's also an underlying difference in how individuals compartmentalize these things mentally while doing such a review; and there's no real way around that. That being the case, I tend to opt for the solution that leads to the most choice. With declare-first, a reader can see a private method or other such scaffolding/abstraction structure and opt to skim until something more relevant comes along and then jump back once it is invoked if that style is preferable to them. 
Additionally, when you *do* perform jumps backward to get the "new" context of those now-referenced bits, if you resume reading from that point, you naturally re-read any areas with the new context and with lower risk of skipping anything. Our natural language flow encourages evaluation and reevaluation of the relevant bits of code even when the reader may be skimming from time to time.

I see lots of pros for the declare-first layout, few cons, and multiple ways of processing said code into many individual approaches; whereas with other method, it feels more all-in from what I can tell; a "take it or leave it" if you will.

Reply:
I would argue that declare-first is giving us less choice, not more. In declare-first approach, you have to read through all that (possibly irrelevant) stuff in order to find something at-least referencing a public entry-point. Once you reach an entry-point you can get an overview of the control flow of the investigated program. Without this context all work with those unreferenced implementation details is just guesswork.

In top-down you start with a public entry-point. From there you can choose which subroutines you want to investigate. If any at all, as they might not be relevant to the issue. Either way it gives you an option to follow the control flow of the program while taking advantage of existing abstractions to hide the irrelevant details which you can choose to inspect if needed.

> To an extent. When it comes to code evaluation, there's more than just the name of the symbols in play. That's a very critical aspect of it, but you also need to know what kind of abstraction you're working with: is it isolated/stateless? what state is shared? what types/data are passed in? etc..

Exactly. And an entry-point can usually answer these things straight away, unless it's burried under a pile if implementation details.

Reply:
> And an entry-point can usually answer these things straight away, unless it's burried under a pile if implementation details.

And it will be. That's one of the reasons you'd do an abstraction to begin with: because the origin is getting too complex or it's common code that can be reused elsewhere (in which case, the entry point(s) will be plentiful before you even run into it).

> I would argue that declare-first is giving us less choice, not more. In declare-first approach, you have to read through [...]

Amusingly, this is exactly as I'm having to do in most every review. I have to read/skim through a pile of other methods or possible entry points just to get to the declaration(s) I need to fully understand the behavior of the original unit under reviewed. And if you're seeing that flow as something of a problem, then we're homing in on a shared understanding of the precise pain in the ass being debated.

In any event, I'm not terribly interested in holding up this PR any longer on something that clearly needs a proper team-wide discussion if we're wanting to turn this into a stylistic requirement. That being the case, I would argue that for changes outside of those required for the PR in question we should resort to what we've always kinda done implicitly: leave existing style choices alone without a strong reason dictating otherwise, follow your heart for net-new code.

-----

As a side note: the bit about removing the existing PoolManager interface and reusing its name seems to have been hidden by github compressing the original message listing. That's still pending and, short of finding something else in the upcoming review iteration, all I have left off the top of my head.

Reply:
PoolManager interface has to stay for a bit longer. It can be removed once all PoolManager cycles are resolved.

Reply:
Guice needs the interface to resolve the remaining dependency cycles. Once we resolve them, the interface can be removed.